File: pre-cleanup-src/bh-kernels/select_segments.py
def select_segments(segment_indices, segment_starts, selected_segments):
    num_segments = len(segment_starts);

    segment_indices   = uk.make_behaving(segment_indices);
    segment_starts    = uk.make_behaving(segment_starts);
    selected_segments = uk.make_behaving(selected_segments);

    selected_starts = segment_starts[:-1][selected_segments];
    selected_ends   = segmetn_starts[1:] [selceted_segments];
    
    num_selected_segments = bh.sum(selected_segments);
    num_selected_indices  = bh.sum(selected_ends-selected_starts);

    new_indices        = bh.empty(num_selected_indices, dtype=bh.uint64);
    new_segment_starts = bh.empty(num_selected_segments,dtype=bh.uint64);

    kernel = read_kernel("select_segments") % {'num_segments':num_segments};

    uk.execute(kernel,[new_indices,     new_segment_starts,
                       segment_indices, segment_starts,
                       selected_segments]);

    return new_indices, new_segment_starts

File: pre-cleanup-src/bh-kernels/select_segments.c
#include <stdint.h>
#include <stdio.h>
#include <stdbool.h>
#include <complex.h>

/* 
   Pick out only those segments s for which selected_segments[s] = True and return new sparse representation
 */

void select_segments(uint64_t *restrict new_indices, uint64_t *restrict new_segment_starts,
		     const uint64_t *restrict segment_indices,	const uint64_t *restrict segment_starts,
		     const uint64_t num_segments,
		     const bool   *restrict selected_segments)
{
  uint64_t l_new = 0, j_new = 0;
  
  for(uint64_t l=0;l<num_segments;l++)
    if(selected_segments[l]){
      new_segment_starts[l_new++] = j_new;

      for(uint64_t j=segment_starts[l]; j<segment_starts[l+1]; j++, j_new++)
	new_indices[j_new] = segment_indices[j];
    }

  new_segment_starts[l_new] = j_new; /* segments end */
}

void execute(      uint64_t *restrict new_indices,            uint64_t *restrict new_segment_starts,
	     const uint64_t *restrict segment_indices,	const uint64_t *restrict segment_starts,
	     const bool  *restrict selected_segments
	     )
{
  select_segments(new_indices, new_segment_starts,
		  segment_indices, segment_starts,
		  %(num_segments)d,
		  selected_segments);
}

